### PR TITLE
AppRole backend

### PR DESCRIPTION
**Pull request by vishalnayak**

### CODE REVIEW

File: builtin/credential/appgroup/backend.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	*framework.Backend
+	salt *salt.Salt
+
+	// Guards to clean-up the expired UserID entries
+	tidyUserIDCASGuard uint32
+
+	// Lock to make changes to registered Apps
+	appLock *sync.RWMutex
+
+	// Lock to make changes to registered Apps
```
Comment (jefferai) [inline builtin/credential/appgroup/backend.go]: "to registered Groups"

File: builtin/credential/appgroup/backend.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+	// Lock to make changes to "generic" mode storage entries
+	genericLock *sync.RWMutex
+
+	// Map of locks to make changes to the UserIDs created.
+	// The lock in the map will be keyed off of UserID itself.
+	// Each UserID will have a separate lock which is used to
+	// update the information related to it, 'num_uses' for example.
+	// The lock will be deleted when the UserID is delted.
+	userIDLocks map[string]*sync.RWMutex
```
Comment (jefferai) [inline builtin/credential/appgroup/backend.go]: This feels like it might be very memory expensive if someone has a whole lot of entries (keeping in mind that most HA backends can't scale to millions of keys anyways). You may want to look at what is being done in the `cubbyhole-the-world` branch to do something similar to token IDs, where we key it off of the first byte, giving 256 locks. That makes it easy to create them ahead of time at backend initialization, which in turns means you don't need to worry about map locking since you're only ever doing reads.

There's another subtle potential problem: it's not safe to have various goroutines reading and writing to a map at once. I don't know if this is also the case if you're having the map return default values, but if it does, you still need to then save that initialized value back in so that the next time someone does a read it's the same one. This is definitely true for a pointer, because the default value will be `nil` so unless you write the initialized value back in you won't be sharing mutexes.

File: builtin/credential/appgroup/path_app.go [hunk 3]
```diff
@@ -0,0 +1,931 @@
...
+	TokenTTL time.Duration `json:"token_ttl" structs:"token_ttl" mapstructure:"token_ttl"`
+
+	// Duration after which an issued token should not be allowed to be renewed
+	TokenMaxTTL time.Duration `json:"token_max_ttl" structs:"token_max_ttl" mapstructure:"token_max_ttl"`
+
+	// If set, activates cubbyhole mode for the UserIDs generated against the App.
+	// An intermediary token will have the actual UserID reponse written in its cubbyhole.
+	// The value of Wrapped will be the duration after which the intermediary token
+	// along with its cubbyhole will be destroyed.
+	Wrapped time.Duration `json:"wrapped" structs:"wrapped" mapstructure:"wrapped"`
```
Comment (jefferai) [inline builtin/credential/appgroup/path_app.go]: In the `cubbyhole-the-world` branch this is called (internally) WrapTTL in the request, and the CLI flag is `-wrap-ttl`. Just using `wrapped` makes it seem like a bool, so we should probably sync this.

File: builtin/credential/appgroup/path_app.go [hunk 4]
```diff
@@ -0,0 +1,931 @@
...
+	// Policies that are to be required by the token to access the App
+	Policies []string `json:"policies" structs:"policies" mapstructure:"policies"`
+
+	// Number of times the UserID generated against the App can be used to perform login
+	NumUses int `json:"num_uses" structs:"num_uses" mapstructure:"num_uses"`
+
+	// Duration (less than the backend's mount) after which a UserID generated against the App will expire
+	UserIDTTL time.Duration `json:"userid_ttl" structs:"userid_ttl" mapstructure:"userid_ttl"`
+
+	// Duration before which an issued token should renew itself
```
Comment (jefferai) [inline builtin/credential/appgroup/path_app.go]: I'd make this "Duration before which an issued token must be renewed"

File: builtin/credential/appgroup/path_app.go [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+
+// appStorageEntry stores all the options that are set on an App
+type appStorageEntry struct {
+	// Policies that are to be required by the token to access the App
+	Policies []string `json:"policies" structs:"policies" mapstructure:"policies"`
+
+	// Number of times the UserID generated against the App can be used to perform login
+	NumUses int `json:"num_uses" structs:"num_uses" mapstructure:"num_uses"`
+
+	// Duration (less than the backend's mount) after which a UserID generated against the App will expire
```
Comment (jefferai) [inline builtin/credential/appgroup/path_app.go]: "...less than the backend mount's max TTL" is I assume what you mean here? Just for clarity.

File: builtin/credential/appgroup/path_app.go [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+			HelpDescription: strings.TrimSpace(appHelp["app-wrapped"][1]),
+		},
+		&framework.Path{
+			Pattern: "app/" + framework.GenericNameRegex("app_name") + "/creds$",
+			Fields: map[string]*framework.FieldSchema{
+				"app_name": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Description: "Name of the App.",
+				},
+				"user_id": &framework.FieldSchema{
```
Comment (jefferai) [inline builtin/credential/appgroup/path_app.go]: Why is this here? It will appear in `path-help`. This may become clearer when I get further down the code but it seems like this is unnecessary.

File: builtin/credential/appgroup/path_app.go [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+	if tokenMaxTTLRaw, ok := data.GetOk("token_max_ttl"); ok {
+		app.TokenMaxTTL = time.Second * time.Duration(tokenMaxTTLRaw.(int))
+	} else if req.Operation == logical.CreateOperation {
+		app.TokenMaxTTL = time.Second * time.Duration(data.Get("token_max_ttl").(int))
+	}
+
+	// Check that the TokenMaxTTL value provided is less than the TokenMaxTTL.
+	// Sanitizing the TTL and MaxTTL is not required now and can be performed
+	// at credential issue time.
+	if app.TokenMaxTTL > time.Duration(0) && app.TokenTTL > app.TokenMaxTTL {
```
Comment (jefferai) [inline builtin/credential/appgroup/path_app.go]: Something I'd consider optional is if you want, you could check whether the TTLs are bigger than the sysview values and attach warnings to the response if so.

File: builtin/credential/appgroup/path_app.go [hunk 8]
```diff
@@ -0,0 +1,931 @@
...
+		&framework.Path{
+			Pattern: "app/" + framework.GenericNameRegex("app_name") + "/wrapped$",
+			Fields: map[string]*framework.FieldSchema{
+				"app_name": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Description: "Name of the App.",
+				},
+				"wrapped": &framework.FieldSchema{
+					Type: framework.TypeDurationSecond,
+					Description: `Duration in seconds, if specified, enables the Cubbyhole mode. In this mode,
```
Comment (jefferai) [inline builtin/credential/appgroup/path_app.go]: It may be worth mentioning that this can also be a string like `"72h"`.

File: builtin/credential/appgroup/path_app.go [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+	return b.handleAppCredsCommon(req, data, true)
+}
+
+func (b *backend) handleAppCredsCommon(req *logical.Request, data *framework.FieldData, specified bool) (*logical.Response, error) {
+	appName := data.Get("app_name").(string)
+	if appName == "" {
+		return logical.ErrorResponse("missing app_name"), nil
+	}
+
+	userID := data.Get("user_id").(string)
```
Comment (jefferai) [inline builtin/credential/appgroup/path_app.go]: OK -- I see now why you had the `user_id` in the field data for both -- but I don't really like it because it will show up in `path-help` which Seth assures me people actually use. Here's my suggestion: change handleAppCredsCommon to be `req *logical.Request, data *framework.FieldData, specified string`. Then in `pathAppCredsSpecificUpdate` you can do `return b.handleAppCredsCommon(req, data, data.Get("user_id").(string))`. Then to decide if it's specific or not you simply check whether the string value is `""`.

File: builtin/credential/appgroup/path_generic.go [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+	"github.com/hashicorp/vault/helper/policyutil"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+// genericStorageEntry stores all the options that are set during UserID
+// creation in "generic" mode.
+type genericStorageEntry struct {
+	// All the Groups that are to be accessible by the UseID created
```
Comment (jefferai) [inline builtin/credential/appgroup/path_generic.go]: s/UseID/UserID/

File: builtin/credential/appgroup/path_generic.go [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+	}
+	data.Raw["user_id"] = userID
+	return b.handleGenericCredsCommon(req, data, false)
+}
+
+func (b *backend) pathGenericCredsSpecificUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	return b.handleGenericCredsCommon(req, data, true)
+}
+
+func (b *backend) handleGenericCredsCommon(req *logical.Request, data *framework.FieldData, specified bool) (*logical.Response, error) {
```
Comment (jefferai) [inline builtin/credential/appgroup/path_generic.go]: Same comment re: specified being a string.

File: builtin/credential/appgroup/path_group.go [hunk 12]
```diff
@@ -0,0 +1,1038 @@
...
+	}
+	data.Raw["user_id"] = userID
+	return b.handleGroupCredsCommon(req, data, false)
+}
+
+func (b *backend) pathGroupCredsSpecificUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	return b.handleGroupCredsCommon(req, data, true)
+}
+
+func (b *backend) handleGroupCredsCommon(req *logical.Request, data *framework.FieldData, specific bool) (*logical.Response, error) {
```
Comment (jefferai) [inline builtin/credential/appgroup/path_group.go]: Same comment about specific being a string.

File: builtin/credential/appgroup/path_login.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	resp, err := b.validateSelector(req.Storage, selectorType, selectorValue)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate selector during renewal:%s", err)
+	}
+
+	return framework.LeaseExtend(resp.TokenTTL, resp.TokenMaxTTL, b.System())(req, data)
+}
+
+func (b *backend) pathLoginUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
```
Comment (jefferai) [inline builtin/credential/appgroup/path_login.go]: I think it'd be really nice if they can either pass in as `selector` and `user_id` or if they can pass in just `user_id` with a selector prepended with e.g. a colon. That way if an organization can easily transmit a string to a service but harder to transmit two strings, they can just merge them together.

File: builtin/credential/appgroup/path_tidy_user_id.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+		return err
+	}
+
+	for _, userID := range userIDs {
+		userIDEntry, err := s.Get("userID/" + userID)
+		if err != nil {
+			return fmt.Errorf("error fetching user ID %s: %s", userID, err)
+		}
+
+		if userIDEntry == nil {
```
Comment (jefferai) [inline builtin/credential/appgroup/path_tidy_user_id.go]: I wonder in these cases if we want to just clean out the user ID instead of erroring. Potentially add it as a `multierror` and keep going, or pass out a `logical.Response` and add it as a warning. The most harm that can happen is login is denied, so we fail secure. But otherwise I wonder if some bug could mean that someone could never get things clean.

(This is also a commentary on the other backends that tidy, too -- they should probably be consistent.)

File: builtin/credential/appgroup/path_tidy_user_id.go [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathTidyUserID(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "tidy/user-id$",
+
```
Comment (jefferai) [inline builtin/credential/appgroup/path_tidy_user_id.go]: Probably a good idea to have a safety buffer to account for clock skew and other issues, since we're checking validity at login time anyways.

File: builtin/credential/appgroup/user_id.go [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+// types, then it should be supported.
+func (b *backend) userIDEntryValid(s logical.Storage, selectorType, selectorValue, userID string) (bool, error) {
+	// If there is no lock indexed by the UserID in the map of locks,
+	// it implies that there is no storage entry corresponding to the
+	// UserID. It is either that the ID is an invalid one or that the
+	// UserID is expired. Fail the validation.
+	b.userIDLocksMapLock.RLock()
+	lock := b.userIDLocksMap[userID]
+	if lock == nil {
+		return false, nil
```
Comment (jefferai) [inline builtin/credential/appgroup/user_id.go]: You'll return here without releasing the lock on the map. You could defer the `RUnlock` but I think here it's safe to simply `RUnlock` the map immediately after fetching `lock`.

File: builtin/credential/appgroup/user_id.go [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+
+// registerUserIDEntry creates a new storage entry for the given UserID.
+// Successful creation of the storage entry results in the creation of a
+// lock in the map of locks maintained at the backend. The index into the
+// map is the UserID itself. During login, if the UserID supplied is not
+// having a corresponding lock in the map, the login attempt fails.
+func (b *backend) registerUserIDEntry(s logical.Storage, selectorType, selectorValue, userID string, userIDEntry *userIDStorageEntry) error {
+	b.userIDLocksMapLock.RLock()
+	if b.userIDLocksMap[userID] != nil {
+		return fmt.Errorf("user ID is already registered")
```
Comment (jefferai) [inline builtin/credential/appgroup/user_id.go]: This'll deadlock. You need to `RUnlock` here before returning.

File: builtin/credential/appgroup/user_id.go [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+		if err != nil {
+			return nil, err
+		}
+		// Append the union of all the policies from all the apps on the group
+		resp.Policies = append(resp.Policies, groupPolicies...)
+
+		// Append the additional policies set on the group
+		resp.Policies = append(resp.Policies, group.AdditionalPolicies...)
+
+		resp.TokenTTL = group.TokenTTL
```
Comment (jefferai) [inline builtin/credential/appgroup/user_id.go]: The case where this and other things (like wrapped) are not set on a group, or are set differently within different component apps, needs to be figured out. It may be that we simply use whatever is set on a group and the defaults otherwise, but there's an argument for the idea that if an app is more restrictive we should restrict within the group, so that the group doesn't get around restrictions set on individual apps.

This can potentially make it a bit harder to reason about, but the flip side is that it guarantees that you can't escape security by simply asking a hapless operator to make a group for you with the values you give her.

Since any given app could be defined multiple times (with different names obviously) then if you actually intend for app owners to have more restrictive credentials than what goes into a group, you can simply have two apps defined and only put one in the group.

File: builtin/credential/appgroup/user_id.go (hunk 17 above)
Comment (jefferai) [inline builtin/credential/appgroup/user_id.go]: I really strongly encourage you to look at `cubbyhole-the-world`'s token store and just allocate the locks ahead of time. It does mean that you may have to hit storage to load a current value but you're doing this anyways. This will make things noticeably simpler to reason about...for instance, no lock switch + refresh.

File: builtin/credential/appgroup/path_app.go [hunk 19]
```diff
@@ -0,0 +1,931 @@
...
+	// Sanitizing the TTL and MaxTTL is not required now and can be performed
+	// at credential issue time.
+	if app.TokenMaxTTL > time.Duration(0) && app.TokenTTL > app.TokenMaxTTL {
+		return logical.ErrorResponse("token_ttl should not be greater than token_max_ttl"), nil
+	}
+
+	// Update only if value is supplied. Defaults to zero.
+	if wrappedRaw, ok := data.GetOk("wrapped"); ok {
+		app.Wrapped = time.Duration(wrappedRaw.(int)) * time.Second
+	}
```
Comment (vishalnayak) [inline builtin/credential/appgroup/path_app.go]: Self-comment: do a `else if req.Operation == logical.CreateOperation` check.

File: builtin/credential/appgroup/path_app.go (hunk 9 above)
Comment (vishalnayak) [inline builtin/credential/appgroup/path_app.go]: Wouldn't it panic if we try to read something which is not present in the schema?

File: builtin/credential/appgroup/user_id.go (hunk 18 above)
Comment (vishalnayak) [inline builtin/credential/appgroup/user_id.go]: I thought of this as well and share your opinion that it could lead to privilege escalation if operators are not cautious while creating groups. It can get difficult to reason about which options are finally being applied on the group, if we start to compare the options on the group with the options on the app.

File: builtin/credential/appgroup/user_id.go (hunk 17 above)
Comment (vishalnayak) [inline builtin/credential/appgroup/user_id.go]: I'll change the locking similar to what is being done in cubbyhole-the-world.

File: builtin/credential/appgroup/path_tidy_user_id.go (hunk 15 above)
Comment (vishalnayak) [inline builtin/credential/appgroup/path_tidy_user_id.go]: I removed safety buffer from this endpoint because the backend's periodic function would not allow any expired entries to live beyond 1 minute past its expiration. So, safety buffer did not make sense in this case to me, given that this 1 minute unavoidable buffer is also not ideal.

Unless, the periodic function should also refer to the configured safety buffer duration, which I did not think was necessary.

File: builtin/credential/appgroup/path_tidy_user_id.go (hunk 14 above)
Comment (vishalnayak) [inline builtin/credential/appgroup/path_tidy_user_id.go]: Since this method was not returning a response object, used multierror to append the errors.

File: builtin/credential/appgroup/path_login.go (hunk 13 above)
Comment (vishalnayak) [inline builtin/credential/appgroup/path_login.go]: Good suggestion. Done!

File: builtin/credential/appgroup/path_app.go (hunk 9 above)
Comment (vishalnayak) [inline builtin/credential/appgroup/path_app.go]: Oh, I don't have to query from both places. Got it.

File: builtin/credential/approle/path_role.go [hunk 20]
```diff
@@ -0,0 +1,1544 @@
...
+	SecretIDTTL time.Duration `json:"secret_id_ttl" structs:"secret_id_ttl" mapstructure:"secret_id_ttl"`
+
+	// Duration before which an issued token must be renewed
+	TokenTTL time.Duration `json:"token_ttl" structs:"token_ttl" mapstructure:"token_ttl"`
+
+	// Duration after which an issued token should not be allowed to be renewed
+	TokenMaxTTL time.Duration `json:"token_max_ttl" structs:"token_max_ttl" mapstructure:"token_max_ttl"`
+
+	// A constraint, if set, requires 'secret_id' credential to be presented during login
+	BoundSecretID bool `json:"bound_secret_id" structs:"bound_secret_id" mapstructure:"bound_secret_id"`
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: This is a bool, so should be `bind_secret_id` rather than `bound_secret_id`.

Comment (jefferai) [inline builtin/credential/approle/path_role.go]: And `BindSecretID` obv :-)

File: builtin/credential/approle/path_role.go [hunk 21]
```diff
@@ -0,0 +1,1544 @@
...
+			HelpDescription: strings.TrimSpace(roleHelp["role-list"][1]),
+		},
+		&framework.Path{
+			Pattern: "role/" + framework.GenericNameRegex("role_name"),
+			Fields: map[string]*framework.FieldSchema{
+				"role_name": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Description: "Name of the role.",
+				},
+				"bound_secret_id": &framework.FieldSchema{
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: s/`bound_secret_id`/`bind_secret_id`

File: builtin/credential/approle/path_role.go [hunk 22]
```diff
@@ -0,0 +1,1544 @@
...
+addresses which can perform the login operation`,
+				},
+				"policies": &framework.FieldSchema{
+					Type:        framework.TypeString,
+					Default:     "default",
+					Description: "Comma separated list of policies on the role.",
+				},
+				"secret_id_num_uses": &framework.FieldSchema{
+					Type:        framework.TypeInt,
+					Description: "Number of times a SecretID can access the role, after which the SecretID will expire.",
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Should maybe mention that the default is 0 and means unlimited.

File: builtin/credential/approle/path_role.go [hunk 23]
```diff
@@ -0,0 +1,1544 @@
...
+				},
+				"secret_id_ttl": &framework.FieldSchema{
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Same with the rest of these TTL params, might want to mention the default is zero which means fall back to mount/system defaults.

File: builtin/credential/approle/path_role.go [hunk 24]
```diff
@@ -0,0 +1,1544 @@
...
+				"token_max_ttl": &framework.FieldSchema{
+					Type:        framework.TypeDurationSecond,
+					Description: "Duration in seconds after which the issued token should not be allowed to be renewed.",
+				},
+				"period": &framework.FieldSchema{
+					Type:    framework.TypeDurationSecond,
+					Default: 0,
+					Description: `If set, indicates that the token generated using this role
+should never expire. The token should be renewed within the
+duration specified by this value. The renewal duration will
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: "The renewal duration will be fixed, if this value is not modified." is a bit awkward...what you're trying to get across is that every renewal will cause the TTL to be set to the value of the period. Perhaps "At each renewal, the token's TTL will be set to the value of this parameter."

File: builtin/credential/approle/path_role.go [hunk 25]
```diff
@@ -0,0 +1,1544 @@
...
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: b.pathRoleBoundCIDRListUpdate,
+				logical.ReadOperation:   b.pathRoleBoundCIDRListRead,
+				logical.DeleteOperation: b.pathRoleBoundCIDRListDelete,
+			},
+			HelpSynopsis:    strings.TrimSpace(roleHelp["role-bound-cidr-list"][0]),
+			HelpDescription: strings.TrimSpace(roleHelp["role-bound-cidr-list"][1]),
+		},
+		&framework.Path{
+			Pattern: "role/" + framework.GenericNameRegex("role_name") + "/bound-secret-id$",
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: s/bound/bind

File: builtin/credential/approle/path_role.go [hunk 26]
```diff
@@ -0,0 +1,1544 @@
...
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: b.pathRoleBoundSecretIDUpdate,
+				logical.ReadOperation:   b.pathRoleBoundSecretIDRead,
+				logical.DeleteOperation: b.pathRoleBoundSecretIDDelete,
+			},
+			HelpSynopsis:    strings.TrimSpace(roleHelp["role-bound-secret-id"][0]),
+			HelpDescription: strings.TrimSpace(roleHelp["role-bound-secret-id"][1]),
+		},
+		&framework.Path{
+			Pattern: "role/" + framework.GenericNameRegex("role_name") + "/num-uses$",
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: s/num_uses/secret_id_num_uses

File: builtin/credential/approle/path_role.go [hunk 27]
```diff
@@ -0,0 +1,1544 @@
...
+	// SecretID generated against the role will expire
+	SecretIDTTL time.Duration `json:"secret_id_ttl" structs:"secret_id_ttl" mapstructure:"secret_id_ttl"`
+
+	// Duration before which an issued token must be renewed
+	TokenTTL time.Duration `json:"token_ttl" structs:"token_ttl" mapstructure:"token_ttl"`
+
+	// Duration after which an issued token should not be allowed to be renewed
+	TokenMaxTTL time.Duration `json:"token_max_ttl" structs:"token_max_ttl" mapstructure:"token_max_ttl"`
+
+	// A constraint, if set, requires 'secret_id' credential to be presented during login
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Note per our Slack discussion - we shouldn't have any required bindings, so it should be valid to turn this off and only use CIDR or only use the role ID.

File: builtin/credential/approle/path_role.go [hunk 28]
```diff
@@ -0,0 +1,1544 @@
...
+	// If the argument to secretIDLock does not start with 2 hex
+	// chars, a generic lock is returned. So, passing empty string
+	// to get the "custom" lock that could be used for listing.
+	lock := b.secretIDLock("")
+	lock.RLock()
+	defer lock.RUnlock()
+
+	// Listing works one level at a time. Get the first level of data
+	// which could then be used to get the actual SecretID storage entries.
+	secretIDHMACs, err := req.Storage.List(fmt.Sprintf("secret_id/%s/", b.salt.SaltID(role.RoleID)))
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: This indicates that the path uses the RoleID -- secret_id/<salted role id>. What happens if they give the same role ID to two roles? The salt is per-backend instead of per-role, so AFAICT you'll end up listing secret IDs for more than one role.

File: builtin/credential/approle/path_role.go [hunk 29]
```diff
@@ -0,0 +1,1544 @@
...
+		return err
+	}
+	if entry == nil {
+		return fmt.Errorf("failed to create storage entry for role %s", roleName)
+	}
+	if err = s.Put(entry); err != nil {
+		return err
+	}
+
+	// Create a storage entry for reverse mroleing of RoleID to role.
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: "mroleing"? :-)

File: builtin/credential/approle/path_role.go [hunk 30]
```diff
@@ -0,0 +1,1544 @@
...
+	if role == nil && req.Operation == logical.CreateOperation {
+		hmacKey, err := uuid.GenerateUUID()
+		if err != nil {
+			return nil, fmt.Errorf("failed to create role_id: %s\n", err)
+		}
+		role = &roleStorageEntry{
+			HMACKey: hmacKey,
+		}
+	} else if role == nil {
+		return nil, fmt.Errorf("Role entry not found when the requested operation is to update it")
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Please have it start lowercase (general Go error rules), but it's also verbose -- "role entry not found during update operation" should suffice :-)

File: builtin/credential/approle/path_role.go [hunk 31]
```diff
@@ -0,0 +1,1544 @@
...
+	// Acquire the lock before deleting the secrets.
+	b.roleLock.Lock()
+	defer b.roleLock.Unlock()
+
+	// Just before the role is deleted, remove all the SecretIDs issued as part of the role.
+	if err = b.flushRoleSecrets(req.Storage, role.RoleID); err != nil {
+		return nil, fmt.Errorf("failed to invalidate the secrets belonging to role '%s': %s", roleName, err)
+	}
+
+	// Delete the reverse mroleing from RoleID to the role
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: There's "mroleing" again, what is it? :-)

File: builtin/credential/approle/path_role.go [hunk 32]
```diff
@@ -0,0 +1,1544 @@
...
+
+	accessorEntry, err := b.secretIDAccessorEntry(req.Storage, secretIDAccessor)
+	if err != nil {
+		return nil, err
+	}
+	if accessorEntry == nil {
+		return nil, fmt.Errorf("failed to find accessor entry for secret_id_accessor:%s\n", secretIDAccessor)
+	}
+
+	entryIndex := fmt.Sprintf("secret_id/%s/%s", b.salt.SaltID(role.RoleID), accessorEntry.SecretIDHMAC)
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: Reminder to switch these to `secret_id/<HMAC(role name, role HMAC key)>/secretIDHMAC`

File: builtin/credential/approle/path_role.go [hunk 33]
```diff
@@ -0,0 +1,1544 @@
...
+		return nil, nil
+	}
+
+	// Deleting a field implies setting the value to it's default value.
+	role.BoundCIDRList = data.GetDefaultOrZero("bound_cidr_list").(string)
+
+	return nil, b.setRoleEntry(req.Storage, roleName, role)
+}
+
+func (b *backend) pathRoleBoundSecretIDUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: s/bound/bind throughout

File: builtin/credential/approle/path_role.go [hunk 34]
```diff
@@ -0,0 +1,1544 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse(fmt.Sprintf("role %s does not exist", roleName)), nil
+	}
+
+	// Currently, only one type of bound is implemented.
+	// Ensure that it is enabled.
+	if !role.BoundSecretID {
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: As discussed, remove this.

File: builtin/credential/approle/path_role.go [hunk 35]
```diff
@@ -0,0 +1,1544 @@
...
+	}
+
+	secretIDStorage := &secretIDStorageEntry{
+		SecretIDNumUses: role.SecretIDNumUses,
+		SecretIDTTL:     role.SecretIDTTL,
+	}
+
+	metadata := data.Get("metadata").(string)
+	if metadata != "" {
+		json.Unmarshal([]byte(metadata), &secretIDStorage.Metadata)
```
Comment (jefferai) [inline builtin/credential/approle/path_role.go]: I wonder if we should require a json object or a simple `<key>=<value>,<key>=<value>`. This way certainly allows more complex metadata, so maybe that's worth it.

File: builtin/credential/approle/secret_id.go [hunk 36]
```diff
@@ -0,0 +1,500 @@
...
+}
+
+// roleIDStorageEntry represents the reverse mroleing from RoleID to Role
+type roleIDStorageEntry struct {
+	Type string `json:"type" structs:"type" mapstructure:"type"`
+	Name string `json:"name" structs:"name" mapstructure:"name"`
+}
+
+// setRoleIDEntry creates a storage entry that maps RoleID to Role
+func (b *backend) setRoleIDEntry(s logical.Storage, roleID string, roleIDEntry *roleIDStorageEntry) error {
```
Comment (jefferai) [inline builtin/credential/approle/secret_id.go]: In the wake of mutable role IDs, probably need to make sure that this is doing the right thing. Or if it's even needed, if the path is going to use the hmac of the role name.

File: builtin/credential/approle/secret_id.go [hunk 37]
```diff
@@ -0,0 +1,500 @@
...
+			if req.Connection != nil {
+				addr = req.Connection.RemoteAddr
+			}
+			if addr == "" || !cidr.Contains(net.ParseIP(addr)) {
+				return nil, fmt.Errorf("unauthorized source address")
+			}
+		}
+	}
+
+	switch {
```
Comment (jefferai) [inline builtin/credential/approle/secret_id.go]: Don't need this switch anymore; any bound values should be checked.

File: builtin/credential/approle/secret_id.go [hunk 38]
```diff
@@ -0,0 +1,500 @@
...
+		if err := s.Delete(accessorEntryIndex); err != nil {
+			return false, fmt.Errorf("failed to delete accessor storage entry: %s", err)
+		}
+		if err := s.Delete(entryIndex); err != nil {
+			return false, fmt.Errorf("failed to delete SecretID: %s", err)
+		}
+	} else {
+		// If the use count is greater than one, decrement it and update the last updated time.
+		result.SecretIDNumUses -= 1
+		result.LastUpdatedTime = time.Now().UTC()
```
Comment (jefferai) [inline builtin/credential/approle/secret_id.go]: I think we decided at HEX not to use `UTC()` any more

File: builtin/credential/approle/secret_id.go [hunk 39]
```diff
@@ -0,0 +1,500 @@
...
+		return nil, err
+	}
+	if entry != nil {
+		return nil, fmt.Errorf("SecretID is already registered")
+	}
+
+	// Create a new entry for the SecretID
+
+	// Set the creation time for the SecretID
+	currentTime := time.Now().UTC()
```
Comment (jefferai) [inline builtin/credential/approle/secret_id.go]: Same comment about UTC

File: builtin/credential/approle/secret_id.go [hunk 40]
```diff
@@ -0,0 +1,500 @@
...
+	}
+
+	return nil
+}
+
+// Iterates through all the Roles, fetches the polices from each Role
+// and returns a union of all the policies combined together. An error
+// is thrown if any Role in the list of Roles supplied is non-existent
+// with the backend.
+func (b *backend) fetchPolicies(s logical.Storage, roles []string) ([]string, error) {
```
Comment (jefferai) [inline builtin/credential/approle/secret_id.go]: I think this is left over from AppGroup? I don't see it used anywhere and it doesn't seem like something needed anymore.

File: builtin/credential/approle/secret_id.go [hunk 41]
```diff
@@ -0,0 +1,500 @@
...
+	SecretIDAccessor string `json:"secret_id_accessor" structs:"secret_id_accessor" mapstructure:"secret_id_accessor"`
+
+	// Number of times this SecretID can be used to perform the login operation
+	SecretIDNumUses int `json:"secret_id_num_uses" structs:"secret_id_num_uses" mapstructure:"secret_id_num_uses"`
+
+	// Duration after which this SecretID should expire. This is
+	// croleed by the backend mount's max TTL value.
+	SecretIDTTL time.Duration `json:"secret_id_ttl" structs:"secret_id_ttl" mapstructure:"secret_id_ttl"`
+
+	// The time in UTC when the SecretID was created
```
Comment (jefferai) [inline builtin/credential/approle/secret_id.go]: If this is ever output (e.g. during a read from an accessor) or some other client call please ensure that it is being output as an RFC value, e.g. `SecretIDNumUses.Format(time.RFC3339)`

When using `structs` to convert things into a map it's easy to forget to sanitize output. One of the reasons I've soured slightly on `structs`.

Same comment for ExpirationTime/LastUpdatedTime.

File: builtin/credential/approle/path_role.go (hunk 34 above)
Comment (vishalnayak) [inline builtin/credential/approle/path_role.go]: Will remove the comment here, but the check is required.

File: builtin/credential/approle/path_role.go (hunk 35 above)
Comment (vishalnayak) [inline builtin/credential/approle/path_role.go]: But supplying input from CLI is much easier if the input is JSON. No?

File: builtin/credential/approle/secret_id.go (hunk 41 above)
Comment (vishalnayak) [inline builtin/credential/approle/secret_id.go]: @jefferai Without making any changes this is how the output looks. Would you still prefer that I change it to RFC3339?

```
creation_time           2016-07-25T13:53:53.477612033-04:00
expiration_time         2016-07-25T23:53:53.477612033-04:00
last_updated_time       2016-07-25T13:53:53.477612033-04:00
```

File: website/source/docs/auth/approle.html.md [hunk 42]
```diff
@@ -0,0 +1,767 @@
...
+endpoint.
+
+### Pull And Push SecretID Modes
+
+If the SecretID generated by the backend is fetched and used for login, it is
+referred as `Pull` mode. If a "custom" SecretID is set against an AppRole by
+the client, it is referred as a `Push` mode.
+
+While the `user_id` of the App-ID backend worked in a `Push` mode, this backend
+recommends the `Pull` mode. The `Pull` mode is supported in AppRole backend,
```
Comment (sandstrom) [inline website/source/docs/auth/approle.html.md]: @vishalnayak shouldn't this say "The `Push` mode is supported in AppRole backend *only* to be able to…"?

Or have I totally misunderstood things? :)

Comment (vishalnayak) [inline website/source/docs/auth/approle.html.md]: @sandstrom, it says that in the `master` branch :-)
